module/component/component.go

```diff
@@ -169,91 +147,117 @@ func NewComponentManagerBuilder() ComponentManagerBuilder {
-		go func() {
-			select {
-			case err := <-errChan:
-				cancel() // shutdown all workers
-
-				// we propagate the error directly to the parent because a failure in a
-				// worker routine is considered irrecoverable
-				parent.Throw(err)
-			case <-c.done:
-				// Without this additional select, there is a race condition here where the done channel
-				// could be closed right after an irrecoverable error is thrown, so that when the scheduler
-				// yields control back to this goroutine, both channels are available to read from. If this
-				// second case happens to be chosen at random to proceed, then we would return and silently
-				// ignore the error.
-				select {
-				case err := <-errChan:
-					cancel()
-					parent.Throw(err)
-				default:
-				}
-			}
+	ctx, cancel := context.WithCancel(parent)
+	signalerCtx, errChan := irrecoverable.WithSignaler(ctx)
+
+	go c.waitForShutdownSignal(ctx.Done())
+
+	// launch goroutine to propagate irrecoverable error
+	go func() {
+		// Closing the done channel here guarantees that any irrecoverable errors encountered will
+		// be propagated to the parent first. Otherwise, there's a race condition between when this
+		// goroutine and the parent's are scheduled. If the parent is scheduled first, any errors
+		// thrown within workers would not have propagated, and it would only receive the done signal
+		defer func() {
+			<-c.workersDone
+			close(c.done)
 		}()
 
-		var workersReady sync.WaitGroup
-		var workersDone sync.WaitGroup
-		workersReady.Add(len(c.workers))
-		workersDone.Add(len(c.workers))
-
-		// launch workers
-		for _, worker := range c.workers {
-			worker := worker
-			go func() {
-				defer workersDone.Done()
-				var readyOnce sync.Once
-				worker(signalerCtx, func() {
-					readyOnce.Do(func() {
-						workersReady.Done()
-					})
-				})
-			}()
+		// wait until the workersDone channel is closed or an irrecoverable error is encountered
+		if err := util.WaitError(errChan, c.workersDone); err != nil {
+			cancel() // shutdown all workers
+
+			// propagate the error directly to the parent because a failure in a worker routine
+			// is considered irrecoverable
+			parent.Throw(err)
 		}
+	}()
 
-		// launch goroutine to close ready channel
-		go c.waitForReady(&workersReady)
+	var workersReady sync.WaitGroup
+	var workersDone sync.WaitGroup
+	workersReady.Add(len(c.workers))
+	workersDone.Add(len(c.workers))
 
-		// launch goroutine to close done channel
-		go c.waitForDone(&workersDone)
-	} else {
-		panic(module.ErrMultipleStartup)
+	// launch workers
+	for _, worker := range c.workers {
+		worker := worker
+		go func() {
+			defer workersDone.Done()
+			var readyOnce sync.Once
+			worker(signalerCtx, func() {
+				readyOnce.Do(func() {
+					workersReady.Done()
+				})
+			})
+		}()
 	}
+
+	// launch goroutine to close ready channel
+	go c.waitForReady(&workersReady)
+
+	// launch goroutine to close workersDone channel
+	go c.waitForDone(&workersDone)
```

Comment:
We probably don't need this anymore right? Instead of deferring `<-c.workersDone` in the error handling goroutine, we can just defer `workersDone.Wait()` instead, and so we can just skip this step.

Reply:
I think we still need to second channel. The challenge is we need a way to signal to the error handler that all of the workers have shutdown and that it's time to stop waiting for errors, and that signal needs to sent after all workers have completely shutdown, otherwise we could miss errors that occur during shutdown

Reply:
Ah okay, would be nice if we had some way of avoiding multiple levels of indirection like this. But I think this is good for now :)